packages/txservice/src/chainreader.ts

```diff
@@ -362,33 +331,48 @@ export class ChainReader {
    */
   public async calculateGasFee(
     chainId: number,
-    gasPriceChainId: number,
     assetId: string,
-    nativeTokenChainId: number,
-    nativeTokenAssetId: string,
     decimals: number,
     method: "prepare" | "fulfill" | "cancel",
     isRouterContract: boolean,
-    requestContext: RequestContext,
-    methodContext?: MethodContext,
+    chainData?: Map<string, ChainData>,
+    _requestContext?: RequestContext,
   ): Promise<BigNumber> {
-    // If the list of chains with deployed Price Oracle Contracts does not include
-    // this chain ID, return 0.
-    if (!CHAINS_WITH_PRICE_ORACLES.includes(chainId) || !CHAINS_WITH_PRICE_ORACLES.includes(nativeTokenChainId)) {
+    const { requestContext, methodContext } = createLoggingContext(this.calculateGasFee.name, _requestContext);
+
+    this.logger.info("Method start", requestContext, methodContext, {
+      chainId,
+      assetId,
+      decimals,
+    });
+
+    const assetIdOnMainnet = await getMainnetEquivalent(chainId, assetId, chainData);
+    const chainIdForTokenPrice = assetIdOnMainnet ? 1 : chainId;
+    const chainIdForGasPrice = chainId;
+    const assetIdForTokenPrice = assetIdOnMainnet ? assetIdOnMainnet : assetId;
+
+    const nativeAssetIdOnMainnet = await getMainnetEquivalent(chainId, constants.AddressZero, chainData);
+    const nativeChainIdForTokenPrice = nativeAssetIdOnMainnet ? 1 : chainId;
+    const nativeAssetIdForTokenPrice = nativeAssetIdOnMainnet || constants.AddressZero;
+
+    if (
+      !CHAINS_WITH_PRICE_ORACLES.includes(chainIdForTokenPrice) ||
+      !CHAINS_WITH_PRICE_ORACLES.includes(nativeChainIdForTokenPrice)
+    ) {
       return constants.Zero;
     }
 
     // Use Ethereum mainnet's price oracle for token reference if no price oracle is present
     // on the specified chain.
     const [ethPrice, tokenPrice, gasPrice] = await Promise.all([
-      this.getTokenPrice(nativeTokenChainId, nativeTokenAssetId),
-      this.getTokenPrice(chainId, assetId),
-      this.getGasPrice(gasPriceChainId, requestContext),
+      this.getTokenPrice(nativeChainIdForTokenPrice, nativeAssetIdForTokenPrice),
+      this.getTokenPrice(chainIdForTokenPrice, assetIdForTokenPrice),
+      this.getGasPrice(chainIdForGasPrice, requestContext),
     ]);
 
     // https://community.optimism.io/docs/users/fees-2.0.html#fees-in-a-nutshell
     let l1GasInUsd = BigNumber.from(0);
-    if (chainId === 10) {
+    if (chainIdForTokenPrice === 10) {
```

Comment:
original chainId is equal to chainIdForGasPrice. so it seems needed to be updated like this
> if (chainIdForGasPrice === 10) {

Reply:
d5f99bf7fafc475dbd8061670e626d8b786a2dee. fixed in this commit